sdk/cpp/agent.cc

```diff
@@ -0,0 +1,93 @@
+// Copyright 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "sdk/cpp/agent.h"
+
+#include <iostream>
+
+#include "sdk/cpp/api.h"
+
+namespace ubbagent {
+
+Agent::Agent(const std::string& config, const std::string& state_dir, absl::Status* out_status) {
+    // Copy the input strings because we need non-const char*.
+    char c_config[config.size() + 1];
+    strcpy(c_config, config.c_str());
+    char c_state_dir[state_dir.size() + 1];
+    strcpy(c_state_dir, state_dir.c_str());
+    // Create a new agent.
+    struct InitResult init_result = AgentInit(c_config, c_state_dir);
+    if (init_result.error_message) {
+        *out_status = absl::InternalError(std::string(init_result.error_message));
+    } else {
+        *out_status = absl::OkStatus();
+        id_ = init_result.id;
+    }
+    free(init_result.error_message);
```

Comment:
I don't do CPP much but this seems odd... Should it really be the responsibility of the caller to free this, considering the fact that this struct is being returned by `AgentInit` (vs passed in)?

Reply:
In order to return a Go string to C, it has to be created in the C heap with C.CString() in Go. To prevent memory leak, it has to be freed in Go with C.free() or in C with free(). https://golang.org/cmd/cgo/#hdr-Go_references_to_C

We can't free it on the Go side because Go won't know when it is safe to free.

It adds a little complexity to the implementation, but on the plus side, teams that use this SDK won't have to worry about it because the string is copied and then freed.